File: reference/python/proofwork/frontier.py
"""The ratchet: progressive bounties that make publishing the profitable move.

The coordination problem in a research network is mostly self-inflicted. A
winner-take-all bounty gives every participant a reason to hoard: if I publish an
improvement so you can build on it, you can extend it by epsilon and take the
whole prize. So nobody shares, everybody duplicates each other's work, and the
network's throughput collapses to that of its best solo participant. People then
try to fix this with elaborate coordination machinery -- work-unit dispatchers,
locks, reservation protocols -- when the actual bug is the payment structure.

Change the structure and most of the problem dissolves:

    An objective carries a monotone best-known score. Whoever moves it from s0
    to s1 is paid in proportion to the distance moved. Publishing is how you get
    paid, so publishing immediately is optimal, and the ledger records who moved
    the frontier as a side effect of paying them.

Three consequences worth stating:

- **Copying is worthless.** Resubmitting a known artifact scores no improvement
  and pays zero. The thing that made hoarding rational is gone.
- **Duplicated work becomes visible.** The frontier is public, so nobody wastes
  compute rediscovering a solution that is already beaten.
- **Attribution is mechanical.** An improvement must cite the frontier entry it
  improved on -- checkable, not a judgement call -- so citation flow pays the
  previous holder automatically. Standing on shoulders is enforced by the
  submission rule rather than by etiquette.

Payouts use a cumulative formula so they telescope exactly:

    cumulative(s) = reward * (clamp(s) - baseline) // (target - baseline)
    payout(s0 -> s1) = cumulative(s1) - cumulative(s0)

Total paid over any sequence of improvements equals cumulative(final_score) --
never more than `reward`, and exactly `reward` if the target is reached. Integer
arithmetic throughout: rounding cannot leak or invent a unit no matter how the
improvements are chopped up.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

MAXIMIZE = "maximize"
MINIMIZE = "minimize"


class RatchetError(ValueError):
    pass


@dataclass(frozen=True)
class Ratchet:
    """Progressive-bounty parameters attached to an objective."""

    baseline: int
    target: int
    reward: int
    direction: str = MAXIMIZE
    #: Smallest score movement that counts. Raising this blocks epsilon-farming
    #: at the cost of also blocking genuine small gains -- a real tradeoff, not
    #: a free win.
    min_improvement: int = 1

    def __post_init__(self) -> None:
        if self.direction not in (MAXIMIZE, MINIMIZE):
            raise RatchetError(f"direction must be {MAXIMIZE} or {MINIMIZE}")
        for name in ("baseline", "target", "reward", "min_improvement"):
            value = getattr(self, name)
            if isinstance(value, bool) or not isinstance(value, int):
                raise RatchetError(f"{name} must be an integer")
        if self.reward < 0:
            raise RatchetError("reward must be non-negative")
        if self.min_improvement < 1:
            raise RatchetError("min_improvement must be at least 1")
        if self.direction == MAXIMIZE and self.target <= self.baseline:
            raise RatchetError("target must exceed baseline when maximizing")
        if self.direction == MINIMIZE and self.target >= self.baseline:
            raise RatchetError("target must be below baseline when minimizing")

    # -- geometry ---------------------------------------------------------
    @property
    def span(self) -> int:
        return abs(self.target - self.baseline)

    def progress(self, score: int) -> int:
        """Distance from baseline toward target, clamped to [0, span]."""
        raw = score - self.baseline if self.direction == MAXIMIZE else self.baseline - score
        return max(0, min(raw, self.span))

    def improves(self, old: int | None, new: int) -> bool:
        if old is None:
            return self.progress(new) >= self.min_improvement
        gain = self.progress(new) - self.progress(old)
        return gain >= self.min_improvement

    # -- money ------------------------------------------------------------
    def cumulative(self, score: int) -> int:
        """Total the objective has paid out once the frontier reaches ``score``."""
        return self.reward * self.progress(score) // self.span

    def payout(self, old: int | None, new: int) -> int:
        """What an improvement from ``old`` to ``new`` earns. Never negative."""
        before = 0 if old is None else self.cumulative(old)
        return max(0, self.cumulative(new) - before)

    def exhausted(self, score: int) -> bool:
        return self.progress(score) >= self.span

    # -- serialization ----------------------------------------------------
    def to_dict(self) -> dict[str, Any]:
        return {
            "baseline": self.baseline,
            "target": self.target,
            "reward": self.reward,
            "direction": self.direction,
            "min_improvement": self.min_improvement,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "Ratchet":
        return cls(
            baseline=data["baseline"],
            target=data["target"],
            reward=data["reward"],
            direction=data.get("direction", MAXIMIZE),
            min_improvement=data.get("min_improvement", 1),
        )


@dataclass(frozen=True)
class FrontierEntry:
    """Who currently holds an objective's best-known score."""

    objective_id: str
    claim_id: str
    holder: str
    score: int
    paid_cumulative: int

    def to_dict(self) -> dict[str, Any]:
        return {
            "objective_id": self.objective_id,
            "claim_id": self.claim_id,
            "holder": self.holder,
            "score": self.score,
            "paid_cumulative": self.paid_cumulative,
        }

File: reference/rust/src/frontier.rs
//! Progressive bounties: pay for distance moved, not for arriving first.
//!
//! A winner-take-all bounty gives everyone a reason to hoard, so nobody shares
//! and everyone rediscovers the same partial results. Paying for the distance
//! an improvement moves makes publishing immediately the profitable move, and
//! payouts telescope -- the pool is exactly exhausted at the target however
//! the curve is chopped.

use crate::canonical::Value;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Direction {
    Maximize,
    Minimize,
}

impl Direction {
    pub fn parse(text: &str) -> Option<Direction> {
        match text {
            "maximize" => Some(Direction::Maximize),
            "minimize" => Some(Direction::Minimize),
            _ => None,
        }
    }

    /// A comparison, never a subtraction: `score - threshold` can overflow
    /// i64 when the two straddle zero at the extremes.
    pub fn clears(&self, score: i64, threshold: i64) -> bool {
        match self {
            Direction::Maximize => score >= threshold,
            Direction::Minimize => score <= threshold,
        }
    }
}

#[derive(Debug, Clone)]
pub struct Ratchet {
    pub baseline: i64,
    pub target: i64,
    pub reward: u64,
    pub direction: Direction,
    pub min_improvement: u64,
}

impl Ratchet {
    pub fn from_value(value: &Value) -> Result<Ratchet, String> {
        let int = |name: &str| -> Result<i64, String> {
            value
                .get(name)
                .and_then(Value::as_i64)
                .ok_or_else(|| format!("ratchet needs an integer {name:?}"))
        };
        let direction = match value.get("direction") {
            None => Direction::Maximize,
            Some(Value::String(text)) => Direction::parse(text)
                .ok_or_else(|| format!("unknown ratchet direction {text:?}"))?,
            Some(_) => return Err("ratchet direction must be a string".into()),
        };
        let reward = value
            .get("reward")
            .and_then(Value::as_u64)
            .ok_or("ratchet needs an integer reward")?;
        let ratchet = Ratchet {
            baseline: int("baseline")?,
            target: int("target")?,
            reward,
            direction,
            min_improvement: match value.get("min_improvement") {
                None => 1,
                Some(Value::Int(n)) if *n >= 1 => {
                    u64::try_from(*n).map_err(|_| "ratchet min_improvement does not fit in u64")?
                }
                Some(Value::Int(_)) => {
                    return Err("ratchet min_improvement must be at least 1".into())
                }
                Some(_) => return Err("ratchet min_improvement must be an integer".into()),
            },
        };
        if ratchet.span() == 0 {
            return Err("ratchet target must differ from baseline".into());
        }
        Ok(ratchet)
    }

    /// Distance from baseline to target. `u64` because i64::MIN..i64::MAX
    /// spans exactly u64::MAX -- the widest legal span is representable, and
    /// the subtraction is done in i128 so it cannot wrap on the way.
    pub fn span(&self) -> u64 {
        let raw = match self.direction {
            Direction::Maximize => i128::from(self.target) - i128::from(self.baseline),
            Direction::Minimize => i128::from(self.baseline) - i128::from(self.target),
        };
        u64::try_from(raw.max(0)).unwrap_or(0)
    }

    /// Distance from baseline toward target, clamped to `[0, span]`.
    pub fn progress(&self, score: i64) -> u64 {
        let raw = match self.direction {
            Direction::Maximize => i128::from(score) - i128::from(self.baseline),
            Direction::Minimize => i128::from(self.baseline) - i128::from(score),
        };
        let span = i128::from(self.span());
        u64::try_from(raw.clamp(0, span)).unwrap_or(0)
    }

    pub fn improves(&self, old: Option<i64>, new: i64) -> bool {
        // In i128 because the difference of two progresses is signed: a
        // regression must come out negative, not wrap to something huge.
        let gained = match old {
            None => i128::from(self.progress(new)),
            Some(old) => i128::from(self.progress(new)) - i128::from(self.progress(old)),
        };
        gained >= i128::from(self.min_improvement)
    }

    /// Total paid once the frontier reaches `score`. `u128` intermediate:
    /// `reward * progress` overflows u64 at realistic values.
    pub fn cumulative(&self, score: i64) -> u64 {
        let span = u128::from(self.span());
        if span == 0 {
            return 0;
        }
        let total = u128::from(self.reward) * u128::from(self.progress(score)) / span;
        u64::try_from(total).unwrap_or(u64::MAX)
    }

    /// What moving `old -> new` earns. Never negative: a regression pays zero
    /// rather than clawing back.
    pub fn payout(&self, old: Option<i64>, new: i64) -> u64 {
        let before = old.map_or(0, |old| self.cumulative(old));
        self.cumulative(new).saturating_sub(before)
    }

    pub fn exhausted(&self, score: i64) -> bool {
        self.progress(score) >= self.span()
    }
}
